# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/common/system_utils_win.cpp
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/common/system_utils_win.cpp
--- a/gfx/angle/checkout/src/common/system_utils_win.cpp	Fri Nov 08 16:41:42 2019 +0000
+++ b/gfx/angle/checkout/src/common/system_utils_win.cpp	Fri Nov 08 17:06:00 2019 +0000
@@ -193,7 +193,11 @@
     {
         startInfo.hStdError = GetStdHandle(STD_ERROR_HANDLE);
     }
-    startInfo.dwFlags |= STARTF_USESTDHANDLES;
+
+    if (stderrOut || stdoutOut)
+    {
+        startInfo.dwFlags |= STARTF_USESTDHANDLES;
+    }
 
     // Create the child process.
     PROCESS_INFORMATION processInfo = {};
@@ -241,13 +245,21 @@
 class Win32Library : public Library
 {
   public:
-    Win32Library(const char *libraryName)
+    Win32Library(const char *libraryName, SearchType searchType)
     {
         char buffer[MAX_PATH];
         int ret = snprintf(buffer, MAX_PATH, "%s.%s", libraryName, GetSharedLibraryExtension());
         if (ret > 0 && ret < MAX_PATH)
         {
-            mModule = LoadLibraryA(buffer);
+            switch (searchType)
+            {
+                case SearchType::ApplicationDir:
+                    mModule = LoadLibraryA(buffer);
+                    break;
+                case SearchType::SystemDir:
+                    mModule = LoadLibraryExA(buffer, nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32);
+                    break;
+            }
         }
     }
 
@@ -275,8 +287,32 @@
     HMODULE mModule = nullptr;
 };
 
-Library *OpenSharedLibrary(const char *libraryName)
+Library *OpenSharedLibrary(const char *libraryName, SearchType searchType)
+{
+    return new Win32Library(libraryName, searchType);
+}
+
+bool IsDirectory(const char *filename)
 {
-    return new Win32Library(libraryName);
+    WIN32_FILE_ATTRIBUTE_DATA fileInformation;
+
+    BOOL result = GetFileAttributesExA(filename, GetFileExInfoStandard, &fileInformation);
+    if (result)
+    {
+        DWORD attribs = fileInformation.dwFileAttributes;
+        return (attribs != INVALID_FILE_ATTRIBUTES) && ((attribs & FILE_ATTRIBUTE_DIRECTORY) > 0);
+    }
+
+    return false;
+}
+
+bool IsDebuggerAttached()
+{
+    return !!::IsDebuggerPresent();
+}
+
+void BreakDebugger()
+{
+    __debugbreak();
 }
 }  // namespace angle